### PR TITLE
timeout waiting for ad server response

### PR DESCRIPTION
**Pull request by zhouyx**
#4864 
As discussed with @jasti, we are setting timeout value to 8s before we have data on the performance. And if ad server fail to respond after 8s, we just treat it as a no-content ad.

### CODE REVIEW

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 1]
```diff
@@ -134,12 +137,17 @@ export class AmpAdApiHandler {
...
-        //TODO: add performance reporting
-        if (this.iframe_) {
-          this.iframe_.style.visibility = '';
-        }
-      });
+      return timerFor(this.baseInstance_.win).timeoutPromise(TIMEOUT_VALUE,
+          this.adResponsePromise_,
+          'timeout waiting for ad response').catch(e => {
+            this.noContent_();
+            user().error(e);
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: let's make it `user().warn()`

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 2]
```diff
@@ -173,6 +173,27 @@ describe('amp-ad-api-handler', () => {
...
             });
       });
+
+      it('should resolve on timeout', () => {
+        const noContentCallbackSpy = sandbox.spy();
+        apiHandler = new AmpAdApiHandler(adImpl, adImpl.element,
+             noContentCallbackSpy);
+        let clock;
+        const beforeAttachedToDom = element => {
+          clock = sandbox.useFakeTimers();
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: does it have to be initialized here?

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 3]
```diff
@@ -173,6 +173,27 @@ describe('amp-ad-api-handler', () => {
...
+          element.setAttribute('data-amp-3p-sentinel', 'amp3ptest' + testIndex);
+          startUpPromise = apiHandler.startUp(element, true);
+        };
+        return createIframeWithMessageStub(window, beforeAttachedToDom)
+            .then(newIframe => {
+              iframe = newIframe;
+              clock.tick(8001);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: add a check before the tick to make sure `noContent` is not called yet

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js (hunk 2 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: You mean the clock? I think as long as it'll work as long as it is initialized before `startUp`  also work. I'll double check.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 1 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: fixed

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js (hunk 2 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: moved outside

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js (hunk 3 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: added

**Comment by lannka** [comment]
LGTM

**Comment by lannka** [comment]
FYI, @jasti changed mind 8s => 10s :-)

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 4]
```diff
@@ -162,6 +170,7 @@ export class AmpAdApiHandler {
     //TODO: make noContentCallback_ default
     if (this.noContentCallback_) {
       this.noContentCallback_();
+      this.noContentCallback_ = null;
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: This should probably be part of `#cleanup`, meaning we'd have to move the cleanup call till after this block.

**Comment by jasti** [comment]
Blame @cramforce :)

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 4 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: fixed, ptal

**Comment by lannka** [comment]
LGTM

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 5]
```diff
@@ -134,12 +137,17 @@ export class AmpAdApiHandler {
...
-        //TODO: add performance reporting
-        if (this.iframe_) {
-          this.iframe_.style.visibility = '';
-        }
-      });
+      return timerFor(this.baseInstance_.win).timeoutPromise(TIMEOUT_VALUE,
+          this.adResponsePromise_,
+          'timeout waiting for ad response').catch(e => {
+            this.noContent_();
+            user().warn(e);
```
Comment (cramforce) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Lets start with `dev().error`, so we get metrics.

**Pull request merged**